api/token.go

```diff
@@ -324,11 +336,22 @@ func (a *API) IdTokenGrant(ctx context.Context, w http.ResponseWriter, r *http.R
 		return badRequestError("Could not read id token grant params: %v", err)
 	}
 
-	if params.IdToken == "" || params.Nonce == "" || params.Provider == "" {
-		return oauthError("invalid request", "id_token, nonce and provider required")
+	if params.IdToken == "" || params.Nonce == "" {
+		return oauthError("invalid request", "id_token and nonce required")
+	}
+	
+	if params.Provider == "" && ( params.ClientID == "" || params.Issuer == "" ) {
+		return oauthError("invalid request", "provider or client_id and issuer required")
 	}
 
-	verifier, err := params.getVerifier(ctx)
+	var verifier *oidc.IDTokenVerifier
+	var err error
+	if params.ClientID == "" || params.Issuer == "" {
+		verifier, err = params.getVerifier(ctx)
+	}
+	if params.Provider == "" {
+		verifier, err = params.getVerifierFromClientIDandIssuer(ctx)
```

Comment:
`getVerifierFromClientIDAndIssuer` seens to require both the `ClientID` and `Issuer` (https://github.com/supabase/gotrue/pull/360/files#diff-150d061ec4206a8396c1e267a3493c43049fb93cca6b50cb99cf4fa8c1adc2e7R107-R111), this check doesn't guarantee that the `ClientID` and `Issuer` will be present if the `Provider` is empty

Reply:
i think a clearer flow would be:

```
if params.Provider != "" {
  verifier, err = params.getVerifier(ctx)
} else if params.ClientID != "" && params.Issuer != "" {
  verifier, err = params.getVerifierFromClientIDAndIssuer(ctx)
} else {
  // return error here
}
```

Alternatively, we can also make this cleaner by having `getVerifier` be a wrapper around `getVerifierFromClientIDAndIssuer` and a new function called `getVerifierFromProvider` and do the check for the above inside `getVerifier`.

Reply:
I followed your first advice.